### PR TITLE
Use unique_ptr for QMCHamiltonian, TrialWaveFunction and MCWalkerConfiguration

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This fixes memory leaks allocated by QMCHamiltonian::makeClone and TrialWaveFunction::makeClone. Following this pattern, I did the same for MCWalkerConfiguration.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted

### CODE REVIEW

File: src/QMCDrivers/CloneManager.cpp [hunk 1]
```diff
@@ -131,21 +123,15 @@ void CloneManager::makeClones(MCWalkerConfiguration& w,
   wClones.resize(NumThreads);
   PsiPoolClones.resize(NumThreads);
   HPoolClones.resize(NumThreads);
-  wClones[0]       = &w;
-  PsiPoolClones[0] = psipool;
-  HPoolClones[0]   = hampool;
```
Comment (quantumsteve) [inline src/QMCDrivers/CloneManager.cpp:136]: objects for thread 0 are owned elsewhere and passed in while objects for other threads are cloned and therefore owning pointers.

I tried making just this change to the develop branch and get the same test failures.

File: src/QMCDrivers/CloneManager.cpp [hunk 2]
```diff
@@ -176,15 +162,12 @@ void CloneManager::makeClones(TrialWaveFunction& guide)
   }
   else
   {
-    guideClones.resize(NumThreads, 0);
+    guideClones.resize(NumThreads);
   }
-  guideClones.resize(NumThreads, 0);
-  guideClones[0] = &guide;
```
Comment (quantumsteve) [inline src/QMCDrivers/CloneManager.cpp:182]: similar pattern here.

File: src/QMCDrivers/CloneManager.cpp [hunk 3]
```diff
@@ -202,20 +185,17 @@ void CloneManager::makeClones(MCWalkerConfiguration& wg, TrialWaveFunction& guid
...
-    wgClones.resize(NumThreads, 0);
+    guideClones.resize(NumThreads);
+    wgClones.resize(NumThreads);
   }
-  guideClones.resize(NumThreads, 0);
-  wgClones.resize(NumThreads, 0);
-  guideClones[0] = &guide;
-  wgClones[0]    = new MCWalkerConfiguration(wg);
-  if (NumThreads == 1)
-    return;
```
Comment (quantumsteve) [inline src/QMCDrivers/CloneManager.cpp:213]: similar pattern here

**Comment by quantumsteve** [comment]
@ye-luo @prckent This is ready for review

**Review by prckent — COMMENTED** [review]
Happy to see the vestigial USE_PARTICLESET_CLONES gone

Slightly worried by some of the codecov lack of coverage warnings given the sensitivity of this code. It LGTM but would be good to hit with a test, particularly makeClones. Assuming that codecov is not confused, did you already try to build something @quantumsteve ? Suggestions @ye-luo , @PDoakORNL ?

File: src/QMCDrivers/WFOpt/CostFunctionCrowdData.cpp [hunk 4]
```diff
@@ -45,13 +45,13 @@ CostFunctionCrowdData::CostFunctionCrowdData(int crowd_size,
 
   for (int ib = 0; ib < crowd_size; ib++)
   {
-    ParticleSet* pCopy         = new ParticleSet(P);
-    TrialWaveFunction* psiCopy = Psi.makeClone(*pCopy);
+    auto pCopy = std::make_unique<ParticleSet>(P);
+    auto psiCopy = Psi.makeClone(*pCopy);
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/CostFunctionCrowdData.cpp:49]: Prefer
```
p_ptr_list_[ib] = std::make_unique<ParticleSet>(P);
auto& pCopy = *p_ptr_list_[ib];
```
This avoid `pCopy` and move.

File: src/QMCDrivers/CloneManager.cpp [hunk 5]
```diff
@@ -98,16 +109,15 @@ void CloneManager::makeClones(MCWalkerConfiguration& w, TrialWaveFunction& psi,
...
-#endif
-      psiClones[ip] = psi.makeClone(*wClones[ip]);
-      hClones[ip]   = ham.makeClone(*wClones[ip], *psiClones[ip]);
+      // all the [ip] objects must be created on the ip threads to have first touch accurate.
+      wClones_uptr[ip - 1] = std::make_unique<MCWalkerConfiguration>(w);
+      wClones[ip] = wClones_uptr[ip - 1].get();
+      psiClones_uptr[ip - 1] = psi.makeClone(*wClones[ip]);
+      psiClones[ip] = psiClones_uptr[ip-1].get();
+      hClones_uptr[ip - 1]   = ham.makeClone(*wClones[ip], *psiClones[ip]);
+      hClones[ip] = hClones_uptr[ip-1].get();
```
Comment (ye-luo) [inline src/QMCDrivers/CloneManager.cpp:120]: Format source code.

File: src/QMCDrivers/CloneManager.h [hunk 6]
```diff
@@ -54,13 +54,17 @@ class CloneManager : public QMCTraits
   ///number of threads
   const IndexType NumThreads;
   ///walkers
```
Comment (ye-luo) [inline src/QMCDrivers/CloneManager.h:56]: Please add note for XXX_uptr pattern.
Describe XXX_uptr[N-1] and XXX[N] relation.

**Review by ye-luo — CHANGES_REQUESTED** [review]
There seems only minor issues to be addressed before merging.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**